### PR TITLE
Relands: Migrate template to Gradle 6.7 and AGP 4.1.0

### LINKED ISSUE #49438: Upgrade the Gradle version in the project template
Gradle 6.0 came out recently: https://docs.gradle.org/6.0/release-notes.html

The Flutter template currently uses 5.6. However, 6.0 might become the default version for new Android apps generated by Android Studio.

For context, this is already the case: https://github.com/flutter/flutter/pull/46260#issuecomment-562802455 if a Flutter app is edited on Android Studio.

### LINKED ISSUE #58247: flutter build apk fails on an app using Android Gradle plugin 4.0
# Canonical Issue
https://issuetracker.google.com/issues/141126614
----------------

## Steps to Reproduce
1. Run `flutter create counter`
2. Update the app to use Android Gradle plugin 4.0.0
3. Run `flutter build apk`

## Logs
```
FAILURE: Build failed with an exception.                                
                                                                        
* What went wrong:                                                      
Execution failed for task ':app:lintVitalRelease'.                      
> Could not resolve all artifacts for configuration ':app:debugRuntimeClasspath'.
   > Failed to transform libs.jar to match attributes {artifactType=processed-jar, org.gradle.libraryelements=jar, org.gradle.usage=java-runtime}.
      > Execution failed for JetifyTransform: /Users/michaelbui/Projects/counter/build/app/intermediates/flutter/debug/libs.jar.
         > Failed to transform '/Users/michaelbui/Projects/counter/build/app/intermediates/flutter/debug/libs.jar' using Jetifier. Reason: FileNotFoundException, message: /Users/michaelbui/Projects/counter/build/app/intermediates/flutter/debug/libs.jar (No such file or directory). (Run with --stacktrace for more details.)
           Please file a bug at http://issuetracker.google.com/issues/new?component=460323.
                                                                        
* Try:                                                                  
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
                                                                        
* Get more help at https://help.gradle.org
```

```
[✓] Flutter (Channel stable, v1.17.2, on Mac OS X 10.15.5 19F96, locale en-AU)
    • Flutter version 1.17.2 at /Users/michaelbui/flutter
    • Framework revision 5f21edf8b6 (20 hours ago), 2020-05-28 12:44:12 -0700
    • Engine revision b851c71829
    • Dart version 2.8.3

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/michaelbui/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.1

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 45.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.45.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.10.2
```

Edit: created an issue against the Jetifier component as suggested by Gradle output https://issuetracker.google.com/u/1/issues/157728192

### PR DESCRIPTION
**Pull request by blasten**
## Description

* Relands https://github.com/flutter/flutter/pull/70808.

* Fixes the regex used to detect the AGP version as this was broken, and added test cases in d63316b.

* The plugins CI issue is fixed by https://github.com/flutter/plugin_tools/pull/123

## Related Issues

Fixes https://github.com/flutter/flutter/issues/58247
Fixes https://github.com/flutter/flutter/issues/49438

* https://issuetracker.google.com/issues/158060799
* https://issuetracker.google.com/issues/158753935

## Tests

Unit tests

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by amirh** [comment]
LGTM for the reland after the change in https://github.com/flutter/plugin_tools/pull/123 is published

File: packages/flutter_tools/lib/src/android/gradle_errors.dart [hunk 1]
```diff
@@ -228,7 +228,7 @@ final GradleHandledError androidXFailureHandler = GradleHandledError(
     }
     if (hasPlugins && usesAndroidX && !shouldBuildPluginAsAar) {
       globals.printStatus(
-        'The built failed likely due to AndroidX incompatibilities in a plugin. '
+        'The build failed likely due to AndroidX incompatibilities in a plugin. '
         'The tool is about to try using Jetfier to solve the incompatibility.'
```
Comment (westy92) [inline packages/flutter_tools/lib/src/android/gradle_errors.dart:232]:
```suggestion
        'The tool is about to try using Jetifier to solve the incompatibility.'
```

**Comment by westy92** [comment]
I'm happy to see this coming through! 🎉 
FYI AGP 4.1.1 and Gradle 6.7.1 are the latest versions.

File: packages/flutter_tools/lib/src/android/gradle_errors.dart (hunk 1 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/android/gradle_errors.dart:232]: Good catch

**Comment by blasten** [comment]
@westy92 we don't necessarily target the latest version since it gets obsolete whenever a new version is released. We also don't bump the Gradle version used in Flutter to match the latest one published by Gradle or Android Gradle plugin. However, we do our best to stay up to date, and ensure backward compatibility.

**Pull request merged**